ResearchKit/Common/ORKFormStepViewController.m

```diff
@@ -651,36 +663,52 @@ - (BOOL)skipButtonEnabled {
 }
 
 - (void)hideSections {
-    
     NSArray<ORKTableSection *> *oldSections = _sections;
     _sections = [NSMutableArray new];
+    
+    NSArray *oldHiddenCellItems = _hiddenCellItems;
+    
+    _hiddenCellItems = [NSMutableArray new];
     _hiddenFormItems = [NSMutableArray new];
     
     ORKTaskResult *taskResult = self.taskViewController.result;
-    NSArray<ORKFormItem *> *formItems = [self allFormItems];
     
     NSMutableIndexSet *sectionsToInsert = [NSMutableIndexSet indexSet];
     NSMutableIndexSet *sectionsToDelete = [NSMutableIndexSet indexSet];
+    NSMutableIndexSet *sectionsToUpdateCells = [NSMutableIndexSet indexSet];
     
-    [formItems enumerateObjectsUsingBlock:^(ORKFormItem * _Nonnull formItem, NSUInteger idx, BOOL * _Nonnull stop) {
-        BOOL hideFormItem = [formItem.hidePredicate evaluateWithObject:@[taskResult]
-                                                 substitutionVariables:@{ORKResultPredicateTaskIdentifierVariableName : taskResult.identifier}];
-        ORKTableSection *section = _allSections[idx];
-        if (!hideFormItem) {
+    for (ORKTableSection *section in _allSections) {
+        BOOL hideSection = YES;
+        for (ORKFormItem *formItem in section.formItems) {
+            BOOL formItemIsHidden = [formItem.hidePredicate evaluateWithObject:@[taskResult]
+                                                         substitutionVariables:@{ORKResultPredicateTaskIdentifierVariableName : taskResult.identifier}];
+            ORKTableCellItem *cellItem = [section cellItemForFormItem:formItem];
+            if (formItemIsHidden) {
+                if (cellItem) {
+                    [_hiddenCellItems addObject:cellItem];
+                }
+            } else {
+                hideSection = NO;
+            }
+        }
+        if (!hideSection) {
             [_sections addObject:section];
             if (![oldSections containsObject:section]) {
                 [sectionsToInsert addIndex:_sections.count - 1];
             }
+            if (section.formItems.count > 1 && ![oldHiddenCellItems isEqualToArray:_hiddenCellItems]) {
+                [sectionsToUpdateCells addIndex:_sections.count - 1];
+            }
```

Comment:
Does _hiddenFormItems need to be modified here?

Reply:
Good catch - I was able to move it further up now. Originally I thought I could get by hiding the entire section if any of the `ORKFormItem`s were to be hidden, but then I realized that would really limit the flexibility and we needed to support hiding at the cell level.